include/materials/material_utility.tcc

```diff
@@ -0,0 +1,221 @@
+//! Compute mean stress p (compression positive)
+inline const double mpm::material::p(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+  return (-1. / 3. * (stress(0) + stress(1) + stress(2)));
+}
+
+//! Compute deviatoric stress
+inline const Eigen::Matrix<double, 6, 1> mpm::material::deviatoric_stress(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute mean p in tension positive
+  const double p = -1.0 * mpm::material::p(stress);
+
+  // Compute deviatoric by subtracting volumetric part
+  Eigen::Matrix<double, 6, 1> deviatoric_stress = stress;
+  for (unsigned i = 0; i < 3; ++i) deviatoric_stress(i) -= p;
+
+  return deviatoric_stress;
+}
+
+//! Compute J2 invariant
+inline const double mpm::material::j2(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  const double j2 = (std::pow((stress(0) - stress(1)), 2) +
+                     std::pow((stress(1) - stress(2)), 2) +
+                     std::pow((stress(0) - stress(2)), 2)) /
+                        6.0 +
+                    std::pow(stress(3), 2) + std::pow(stress(4), 2) +
+                    std::pow(stress(5), 2);
+
+  return j2;
+}
+
+//! Compute J3 invariant
+inline const double mpm::material::j3(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute deviatoric stress
+  Eigen::Matrix<double, 6, 1> deviatoric_stress =
+      mpm::material::deviatoric_stress(stress);
+
+  // Compute J3
+  const double j3 =
+      (deviatoric_stress(0) * deviatoric_stress(1) * deviatoric_stress(2)) -
+      (deviatoric_stress(2) * std::pow(deviatoric_stress(3), 2)) +
+      ((2 * deviatoric_stress(3) * deviatoric_stress(4) *
+        deviatoric_stress(5)) -
+       (deviatoric_stress(0) * std::pow(deviatoric_stress(4), 2)) -
+       (deviatoric_stress(1) * std::pow(deviatoric_stress(5), 2)));
+
+  return j3;
+}
+
+//! Compute deviatoric q
+inline const double mpm::material::q(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute J2 from
+  const double j2 = mpm::material::j2(stress);
+
+  return (std::sqrt(3 * j2));
+}
+
+//! Compute Lode angle
+inline const double mpm::material::lode_angle(
+    const Eigen::Matrix<double, 6, 1>& stress) {
+
+  // Compute j2 and j3
+  const double j2 = mpm::material::j2(stress);
+  const double j3 = mpm::material::j3(stress);
+
+  // Compute Lode angle value
+  double lode_angle_val = 0.0;
+  if (std::abs(j2) > std::numeric_limits<double>::epsilon()) {
```

Comment:
Isn't this such a small value? Why can't we set a default tolerance, which constitutive modelers can modify?

Reply:
Yes, it is such a small value and not likely to happen except almost zero stress (j2 is a squared term). But not all constitutive models have tolerance.

Reply:
If we have a tolerance value as a default argument, then modelers needing a tolerance can use it, others can ignore. Any reason, why we shouldn't do this?

Reply:
I don't mind actually, I will add that.

Reply:
I second @kks32

Reply:
I have overloaded the functions

Reply:
Set a default argument, please don't overload the functions